File: benchmarks/pandas/bench_interval.py
"""Benchmark: Interval / IntervalIndex — closed/open intervals."""
import json, time
import pandas as pd

SIZE = 10_000
WARMUP = 5
ITERATIONS = 50

intervals = [pd.Interval(i, i + 1) for i in range(SIZE)]
breaks = list(range(1_001))

for _ in range(WARMUP):
    for iv in intervals[:100]:
        iv.mid in iv
        _ = iv.length
        str(iv)
    pd.IntervalIndex.from_breaks(breaks)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    for iv in intervals:
        iv.mid in iv
        _ = iv.length
        str(iv)
    pd.IntervalIndex.from_breaks(breaks)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({"function":"interval","mean_ms":round(total_ms/ITERATIONS,3),"iterations":ITERATIONS,"total_ms":round(total_ms,3)}))

File: benchmarks/tsb/bench_interval.ts
/**
 * Benchmark: Interval / IntervalIndex — closed/open intervals.
 * Outputs JSON: {"function": "interval", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Interval, IntervalIndex } from "../../src/index.ts";

const SIZE = 10_000;
const WARMUP = 5;
const ITERATIONS = 50;

const intervals = Array.from({ length: SIZE }, (_, i) => new Interval(i, i + 1));
const breaks = Array.from({ length: 1_001 }, (_, i) => i);

for (let i = 0; i < WARMUP; i++) {
  for (const iv of intervals.slice(0, 100)) {
    void iv.contains(iv.mid);
    void iv.length;
    void iv.toString();
  }
  IntervalIndex.fromBreaks(breaks);
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const start = performance.now();
  for (const iv of intervals) {
    void iv.contains(iv.mid);
    void iv.length;
    void iv.toString();
  }
  IntervalIndex.fromBreaks(breaks);
  times.push(performance.now() - start);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "interval",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);
